File: native-applications/ios/Forumline/WebViewBridge.swift
import WebKit

/// Singleton for native → web messaging
class WebViewBridge {
    static let shared = WebViewBridge()
    weak var webView: WKWebView?

    func sendToWeb(_ message: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: message),
              let jsonString = String(data: data, encoding: .utf8)
        else { return }

        let js = "window.forumlineNativeBridge?.onMessage(\(jsonString));"
        DispatchQueue.main.async {
            self.webView?.evaluateJavaScript(js) { _, error in
                if let error = error {
                    print("[Forumline] Bridge send error: \(error)")
                }
            }
        }
    }
}

File: native-applications/windows/WebViewBridge.cs
using System.Text.Json;
using Microsoft.Web.WebView2.Core;

namespace Forumline;

/// <summary>
/// Singleton for native → web messaging.
/// Windows equivalent of iOS/Android/macOS WebViewBridge.
/// </summary>
public sealed class WebViewBridge
{
    public static WebViewBridge Instance { get; } = new();

    public CoreWebView2? CoreWebView { get; set; }

    /// <summary>
    /// Send a message from native to the web app.
    /// </summary>
    public void SendToWeb(object message)
    {
        if (CoreWebView == null) return;

        var json = JsonSerializer.Serialize(message);
        var js = $"window.forumlineNativeBridge?.onMessage({json});";

        System.Windows.Application.Current?.Dispatcher.InvokeAsync(() =>
        {
            CoreWebView.ExecuteScriptAsync(js);
        });
    }
}
